[autograd] remove spinning for dist engine

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36606 [autograd] remove spinning for dist engine**
* #37061 [autograd] refactor some functions

This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works. To achieve that:
1. remove the continuation logic in
execute_graph_task_with_continuiation
2. separate the usage of execute_graph_task between dist_engine and
local engine, now dist_engine universally use
`execute_graph_task_until_ready_queue_empty` (a better name appreciated
here).
3. remove enqueue_blocked_task_on_cpu
4. remove the async mode in `execute_with_graph_task` as we don't need
to use it in dist_engine
5. move some Engine functions to public to make it accessible from dist
engine

Differential Revision: [D21032731](https://our.internmc.facebook.com/intern/diff/D21032731)

wanchaol committed:
[WIP][autograd] remove spinning for dist engine

This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works.

[ghstack-poisoned]

wanchaol committed:
Update on "[autograd] remove spinning for dist engine"


This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works.

[ghstack-poisoned]

wanchaol committed:
Update on "[autograd] remove spinning for dist engine"


This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works. To achieve that:
1. remove the continuation logic in
execute_graph_task_with_continuiation
2. separate the usage of execute_graph_task between dist_engine and
local engine, now dist_engine universally use
`execute_graph_task_until_ready_queue_empty` (a better name appreciated
here).
3. remove enqueue_blocked_task_on_cpu
4. remove the async mode in `execute_with_graph_task` as we don't need
to use it in dist_engine

[ghstack-poisoned]

wanchaol committed:
Update on "[autograd] remove spinning for dist engine"


This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works. To achieve that:
1. remove the continuation logic in
execute_graph_task_with_continuiation
2. separate the usage of execute_graph_task between dist_engine and
local engine, now dist_engine universally use
`execute_graph_task_until_ready_queue_empty` (a better name appreciated
here).
3. remove enqueue_blocked_task_on_cpu
4. remove the async mode in `execute_with_graph_task` as we don't need
to use it in dist_engine

Differential Revision: [D21032731](https://our.internmc.facebook.com/intern/diff/D21032731)

[ghstack-poisoned]

gqchen commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -809,35 +809,35 @@ void Engine::initialize_device_threads_pool() {
       if (task.fn_ && !local_graph_task->has_error_.load()) {
         AutoGradMode grad_mode(local_graph_task->grad_mode_);
         try {
           evaluate_function(local_graph_task, task.fn_.get(), task.inputs_);
```
we need to do continuation in evaluate_function? It seems it's not part of this PR.

gqchen replied:
We certainly don't want to change local engine. 

I think what we want to avoid here is to get stuck silently when someone happens to do distributed autograd using GPU. Can we add an assertion here if node is expected to run on GPU?

gqchen commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -809,35 +809,35 @@ void Engine::initialize_device_threads_pool() {
-}
-
-void Engine::execute_graph_task_with_continuation(const std::shared_ptr<GraphTask>& graph_task) {
   std::shared_ptr<ReadyQueue> graph_task_rq = graph_task->cpu_ready_queue_;
```
If we share one ready queue for the whole graph task, everytime a new backward is called, we create a new worker which can take nodes from the same readyqueue other backward calls populated. Our general threading strategy is fork and rpc in the forward path makes the corresponding backward  parallelizable, and ops in the corresponding "fork" and "rpc" should be run in the same thread. The behavior here makes backward and forward unsymmetric. It will be much harder to reason and explain.

gqchen replied:
definitely.

gqchen commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -850,19 +850,13 @@ void Engine::execute_graph_task_with_continuation(const std::shared_ptr<GraphTas
-    });
   }
-
+  return graph_task->future_result_;
```
this PR doesn't address refcount issue? Without that fix, this will cause training failures.

wanchaol replied:
it seems that we don't need to change to return `shared_ptr<FutureVariableList>&` as of @jjlilley 's change in https://github.com/pytorch/pytorch/pull/36502, I will need to verify if this is the case.

wanchaol committed:
Update on "[autograd] remove spinning for dist engine"


This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works. To achieve that:
1. remove the continuation logic in
execute_graph_task_with_continuiation
2. separate the usage of execute_graph_task between dist_engine and
local engine, now dist_engine universally use
`execute_graph_task_until_ready_queue_empty` (a better name appreciated
here).
3. remove enqueue_blocked_task_on_cpu
4. remove the async mode in `execute_with_graph_task` as we don't need
to use it in dist_engine

Differential Revision: [D21032731](https://our.internmc.facebook.com/intern/diff/D21032731)

[ghstack-poisoned]

wanchaol commented:
@gqchen I just had a offline discussion with @albanD, we will refactor the existing engine `thread_main` to make it works well with both the vanilla engine and dist engine. So that in dist_engine, we don't need to increase `outstanding_task` to keep the graph_task alive, and could remove this `execute_graph_task_until_ready_queue_empty` redundant logic. This would serve us better on solving GPU training issue (no continuation needed, we will just waiting for the GPU task to finish before we exit the thread_main).

gqchen commented:
That's great!

wanchaol committed:
Update on "[autograd] remove spinning for dist engine"


This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works. To achieve that:
1. remove the continuation logic in
execute_graph_task_with_continuiation
2. separate the usage of execute_graph_task between dist_engine and
local engine, now dist_engine universally use
`execute_graph_task_until_ready_queue_empty` (a better name appreciated
here).
3. remove enqueue_blocked_task_on_cpu
4. remove the async mode in `execute_with_graph_task` as we don't need
to use it in dist_engine

Differential Revision: [D21032731](https://our.internmc.facebook.com/intern/diff/D21032731)

[ghstack-poisoned]

wanchaol committed:
Update on "[autograd] remove spinning for dist engine"


This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works. To achieve that:
1. remove the continuation logic in
execute_graph_task_with_continuiation
2. separate the usage of execute_graph_task between dist_engine and
local engine, now dist_engine universally use
`execute_graph_task_until_ready_queue_empty` (a better name appreciated
here).
3. remove enqueue_blocked_task_on_cpu
4. remove the async mode in `execute_with_graph_task` as we don't need
to use it in dist_engine

Differential Revision: [D21032731](https://our.internmc.facebook.com/intern/diff/D21032731)

[ghstack-poisoned]

wanchaol commented:
@albanD this is now ready for review.

wanchaol committed:
Update on "[autograd] remove spinning for dist engine"


This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works. To achieve that:
1. remove the continuation logic in
execute_graph_task_with_continuiation
2. separate the usage of execute_graph_task between dist_engine and
local engine, now dist_engine universally use
`execute_graph_task_until_ready_queue_empty` (a better name appreciated
here).
3. remove enqueue_blocked_task_on_cpu
4. remove the async mode in `execute_with_graph_task` as we don't need
to use it in dist_engine
5. move some Engine functions to public to make it accessible from dist
engine

Differential Revision: [D21032731](https://our.internmc.facebook.com/intern/diff/D21032731)

[ghstack-poisoned]

albanD commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -254,17 +248,71 @@ void DistEngine::computeDependencies(
+      if (task.fn_ && !local_graph_task->has_error_.load()) {
+        AutoGradMode grad_mode(local_graph_task->grad_mode_);
+        try {
+          engine_.evaluate_function(local_graph_task, task.fn_.get(), task.inputs_, cpu_ready_queue);
```
Not sure if you this is supposed to support reentrant autograd. If so, I think it is missing an `init_thread_local_ready_queue()`. Otherwise, it's fine.

wanchaol replied:
I think in the dist engine call phase, it's a bit hard now to support reentrant call from `distributed::backward()` level, but for local reentrant backward that dist backward calls, it might be possible to support it. The tricky here is, we are using a locally created ready_queue per `execute_until_ready_queue_empty`, to execute the same graph task concurrently, if we `init_thread_local_ready_queue()`, there's chance that the thread local ready_queue get overwrited by two different threads concurrently (e.g. thread A and B doing local reentrant backward concurrently). We might need to think more on how to handle local reentrant backward calls indeed, right now the approach is to use a new ready queue whenever local reentrant calls `Engine::execute`. I would leave it as a follow up task.

albanD replied:
But this is a thread local variable. so different thread should not overwrite each other right?

I am fine if this is not supported, just wanted to point it out. Maybe you want to add a comment to make sure this is clear.

wanchaol replied:
Oh actually you are right, since it's a thread local it will be preserved correctly. Let me try to add it :)

albanD reviewed (approved):
The vanilla engine part looks good.

Small nit in the distributed one if you want to support reentrant. Otherwise it looks ok from a high level.

wanchaol committed:
Update on "[autograd] remove spinning for dist engine"


This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works. To achieve that:
1. remove the continuation logic in
execute_graph_task_with_continuiation
2. separate the usage of execute_graph_task between dist_engine and
local engine, now dist_engine universally use
`execute_graph_task_until_ready_queue_empty` (a better name appreciated
here).
3. remove enqueue_blocked_task_on_cpu
4. remove the async mode in `execute_with_graph_task` as we don't need
to use it in dist_engine
5. move some Engine functions to public to make it accessible from dist
engine

Differential Revision: [D21032731](https://our.internmc.facebook.com/intern/diff/D21032731)

[ghstack-poisoned]

wanchaol committed:
Update on "[autograd] remove spinning for dist engine"


This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works. To achieve that:
1. remove the continuation logic in
execute_graph_task_with_continuiation
2. separate the usage of execute_graph_task between dist_engine and
local engine, now dist_engine universally use
`execute_graph_task_until_ready_queue_empty` (a better name appreciated
here).
3. remove enqueue_blocked_task_on_cpu
4. remove the async mode in `execute_with_graph_task` as we don't need
to use it in dist_engine
5. move some Engine functions to public to make it accessible from dist
engine

Differential Revision: [D21032731](https://our.internmc.facebook.com/intern/diff/D21032731)

[ghstack-poisoned]

gqchen commented on torch/csrc/distributed/autograd/engine/dist_engine.h:
```diff
@@ -88,13 +88,47 @@ class TORCH_API DistEngine {
+  // over the network.
+  //
+  // XXX: calling this function assumes that we will have NO GPU nodetasks be executed
+  // for the graph_task, the caller of this function need to ensure this otherwise
```
can we do a follow up task to add some assertion for this?

gqchen commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -254,17 +248,71 @@ void DistEngine::computeDependencies(
   autogradContext->setGraphTask(std::move(graphTask));
 }
 
+std::shared_ptr<FutureVariableList>& DistEngine::execute_graph_task_until_ready_queue_empty(
```
love the idea of moving this function to DistEngine!

gqchen reviewed (approved):
looks great!

wanchaol committed:
launch the computation instead on "[autograd] remove spinning for dist engine"


This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works. To achieve that:
1. remove the continuation logic in
execute_graph_task_with_continuiation
2. separate the usage of execute_graph_task between dist_engine and
local engine, now dist_engine universally use
`execute_graph_task_until_ready_queue_empty` (a better name appreciated
here).
3. remove enqueue_blocked_task_on_cpu
4. remove the async mode in `execute_with_graph_task` as we don't need
to use it in dist_engine
5. move some Engine functions to public to make it accessible from dist
engine

Differential Revision: [D21032731](https://our.internmc.facebook.com/intern/diff/D21032731)

[ghstack-poisoned]

wanchaol committed:
Update on "[autograd] remove spinning for dist engine"


This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works. To achieve that:
1. remove the continuation logic in
execute_graph_task_with_continuiation
2. separate the usage of execute_graph_task between dist_engine and
local engine, now dist_engine universally use
`execute_graph_task_until_ready_queue_empty` (a better name appreciated
here).
3. remove enqueue_blocked_task_on_cpu
4. remove the async mode in `execute_with_graph_task` as we don't need
to use it in dist_engine
5. move some Engine functions to public to make it accessible from dist
engine

Differential Revision: [D21032731](https://our.internmc.facebook.com/intern/diff/D21032731)

[ghstack-poisoned]

wanchaol committed:
Update on "[autograd] remove spinning for dist engine"


This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works. To achieve that:
1. remove the continuation logic in
execute_graph_task_with_continuiation
2. separate the usage of execute_graph_task between dist_engine and
local engine, now dist_engine universally use
`execute_graph_task_until_ready_queue_empty` (a better name appreciated
here).
3. remove enqueue_blocked_task_on_cpu
4. remove the async mode in `execute_with_graph_task` as we don't need
to use it in dist_engine
5. move some Engine functions to public to make it accessible from dist
engine

Differential Revision: [D21032731](https://our.internmc.facebook.com/intern/diff/D21032731)

[ghstack-poisoned]

wanchaol committed:
Update on "[autograd] remove spinning for dist engine"


This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works. To achieve that:
1. remove the continuation logic in
execute_graph_task_with_continuiation
2. separate the usage of execute_graph_task between dist_engine and
local engine, now dist_engine universally use
`execute_graph_task_until_ready_queue_empty` (a better name appreciated
here).
3. remove enqueue_blocked_task_on_cpu
4. remove the async mode in `execute_with_graph_task` as we don't need
to use it in dist_engine
5. move some Engine functions to public to make it accessible from dist
engine

Differential Revision: [D21032731](https://our.internmc.facebook.com/intern/diff/D21032731)

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -1032,7 +974,6 @@ bool Engine::is_checkpoint_valid() {
 void Engine::init_local_ready_queue(std::shared_ptr<ReadyQueue> ready_queue) {
   if (ready_queue) {
     // if ready_queue provided in the caller, use the caller's ready_queue to initialize local_ready_queue
-    TORCH_INTERNAL_ASSERT(!local_ready_queue);
```
Why do you need to remove that? This should always be true, even from the distributed call no? Otherwise you're doing reentrant for a dist engine? :D

wanchaol replied:
In the distributed engine the `execute_graph_task_until_rq_empty` is called from a thread pool (RPC threadpool or JIT threadpool), so if the thread pool launches the function twice or more, the local_ready_queue already been set to the ready queue that's created by previous calls. That's why I remove the assert here.

albanD replied:
Ok,
I might be cleaning this all logic later. Will be fine for now.

wanchaol replied:
actually I did a couple of runs with initializing the local_ready_queue to support reentrant backwards, it turns out the perf becomes bad after that (around 10% drop), this is pretty surprising as I wouldn't expect a big drop like this. So I reverted this change and added a TODO in the comment to investigate further in the future.

wanchaol committed:
Update on "[autograd] remove spinning for dist engine"


This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works. To achieve that:
1. remove the continuation logic in
execute_graph_task_with_continuiation
2. separate the usage of execute_graph_task between dist_engine and
local engine, now dist_engine universally use
`execute_graph_task_until_ready_queue_empty` (a better name appreciated
here).
3. remove enqueue_blocked_task_on_cpu
4. remove the async mode in `execute_with_graph_task` as we don't need
to use it in dist_engine
5. move some Engine functions to public to make it accessible from dist
engine

Differential Revision: [D21032731](https://our.internmc.facebook.com/intern/diff/D21032731)

[ghstack-poisoned]
